## Задача 1. My heart will go on

Датасет **titanic** из библиотеки `Seaborn` содержит информацию о пассажирах легендарного корабля Титаник, который затонул в 1912 году после столкновения с айсбергом. Этот набор данных часто используется для обучения и тестирования алгоритмов машинного обучения, особенно в задачах бинарной классификации (выжил / не выжил).

**Описание данных**

| Поле         | Тип      | Описание |
|--------------|----------|----------|
| `survived`   | int      | Выжил (1) или не выжил (0) |
| `pclass`     | int      | Класс билета (1, 2, 3) |
| `sex`        | str      | Пол (`male`/`female`) |
| `age`        | float    | Возраст |
| `sibsp`      | int      | Количество братьев/сестёр/супругов на борту |
| `parch`      | int      | Количество родителей/детей на борту |
| `fare`       | float    | Стоимость билета |
| `embarked`   | str      | Порт посадки (`C`=Cherbourg, `Q`=Queenstown, `S`=Southampton) |
| `class`      | str      | Класс билета (`First`, `Second`, `Third`) |
| `who`        | str      | Категория: `man`, `woman` или `child` |
| `adult_male` | bool     | Является ли взрослым мужчиной |
| `deck`       | str      | Палуба |
| `embark_town`| str      | Название порта посадки |
| `alive`      | str      | Выжил (`yes`/`no`) |
| `alone`      | bool     | Путешествовал один |

**Загрузка датасета**

In [1]:
import seaborn as sns


titanic_data = sns.load_dataset("titanic")
titanic_data.sample(5)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
603,0,3,male,44.0,0,0,8.050,S,Third,man,True,NaN,Southampton,no,True
555,0,1,male,62.0,0,0,26.550,S,First,man,True,NaN,Southampton,no,True
417,1,2,female,18.0,0,2,13.000,S,Second,woman,False,NaN,Southampton,yes,False
468,0,3,male,NaN,0,0,7.725,Q,Third,man,True,NaN,Queenstown,no,True
2,1,3,female,26.0,0,0,7.925,S,Third,woman,False,NaN,Southampton,yes,True


**Задача**

Ниже описаны 10 небольших заданий, которые вам необходимо решить.

**Подсказка**:

В некоторых заданиях вам может быть полезен метод `value_counts`.

### Часть 1

Определите число пропущенных данных для каждого столбца таблицы `titanic_data`.

In [ ]:

titanic_data.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

### Часть 2

Удалите все столбцы, количество пропусков в которых превышает половину количества строк в таблице.

После того, как вы удалите все столбцы, нарушающие описанное условие, удалите все строки, количество пропусков в которых превышает половину количества столбцов.

In [16]:

rows, cols = titanic_data.shape

titanic_data.dropna(thresh=rows / 2, axis=1, inplace=True)
titanic_data.dropna(thresh=cols / 2, axis=0, inplace=True)

titanic_data.isnull().sum()


survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
embark_town      2
alive            0
alone            0
dtype: int64

### Часть 3

Если вы сделали все правильно, больше всего пропусков должно остаться в столбце `"age"` - 177. Их необходимо заполнить. Заполним пропуски следующим образом:
- Если значение столбца `"who"="man"`, пропуск необходимо заполнить медианным значением известных возрастов мужчин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="woman"`, пропуск необходимо заполнить медианным значением известных возрастов женщин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="child"`, пропуск необходимо заполнить медианным значением известных возрастов детей, округленным до ближайшего целого числа;

In [ ]:

whos = ['man', 'woman', 'child']

ages = [
    round(titanic_data.loc[titanic_data['who'] == x, 'age'].median()) for x in whos
]

changed = [    
    titanic_data['age'][titanic_data['who'] == who].fillna(age) for who, age in zip(whos, ages)
]

for who, change in zip(whos, changed):
    titanic_data.loc[titanic_data['who'] == who, 'age'] = change 

titanic_data.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
embark_town    2
alive          0
alone          0
dtype: int64

### Часть 4

Удалите все строки, в которых осталось больше одного пропуска. Если вы все сделали правильно, после этого действия в таблице не должно остаться пропусков.

In [34]:
titanic_data.dropna(thresh=titanic_data.shape[1] - 1, inplace=True)

titanic_data.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

### Часть 5

Определите название города, из которого отправилось больше всего пассажиров.

In [ ]:
titanic_data['embark_town'].value_counts().index[0]

'Southampton'

### Часть 6

Определите процент выживших пассажиров от числа пассажиров, оставшихся в таблице после очистки данных. Ответ округлите до 2 знаков после запятой.

In [ ]:
count_surv = titanic_data['survived'].sum()
round(count_surv / titanic_data.shape[0] * 100, 2)

np.float64(38.25)

### Часть 7

Определите число выживших пассажиров для каждого пункта отправления. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия пунктов отправления, а значения - число выживших пассажиров.

In [ ]:
titanic_data.loc[titanic_data['survived'] == True, 'embark_town'].value_counts()

embark_town
Southampton    217
Cherbourg       93
Queenstown      30
Name: count, dtype: int64

### Часть 8

Определите процент выживших пассажиров в каждом классе. Значения округлите до 2 знаков после запятой. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия классов, а значения - процент выживших пассажиров.

In [ ]:
surv_from_class = titanic_data.loc[titanic_data['survived'] == 1, 'class'].value_counts()
count_class = titanic_data['class'].value_counts()

round(surv_from_class / count_class * 100, 2)


class
First     62.62
Second    47.28
Third     24.24
Name: count, dtype: float64

### Часть 9

Будем считать, что пассажиры, купившие билет **НЕ МЕНЕЕ** чем за $100, считаются богатыми. Определите процент выживших среди богатых пассажиров. Ответ округлите до 2 знаков после запятой. В ответе должно получиться число. 

In [ ]:
rich_passanger = titanic_data.loc[titanic_data['fare'] >= 100]
surv_count = rich_passanger['survived'].sum()

round(surv_count / rich_passanger.shape[0] * 100, 2)

np.float64(73.58)

### Часть 10

Определите количество детей, путешествовавших в одиночку.

In [ ]:
titanic_data.loc[titanic_data['who'] == 'child', 'alone'].sum()

np.int64(6)

Какие выводы вы можете сделать о выживших пассажирах Титаника? 